File: apps/ios/messen/Features/Chat/ChatViewModel.swift
import Foundation
import Combine

@MainActor
final class ChatViewModel: ObservableObject {
    @Published var messages: [Message] = []
    @Published var loading: Bool = false
    @Published var error: String?
    @Published var typingUsers: Set<String> = []
    @Published var draft: String = ""

    /// Сообщение, на которое сейчас отвечаем.
    @Published var replyingTo: Message?
    /// Сообщение, которое сейчас редактируется.
    @Published var editingMessage: Message?

    let chat: Chat
    private let currentUserId: String
    private let privateKey: String?

    private var cancellables = Set<AnyCancellable>()
    private var typingTimer: Timer?

    init(chat: Chat, currentUserId: String, privateKey: String?) {
        self.chat = chat
        self.currentUserId = currentUserId
        self.privateKey = privateKey
        subscribeToSocket()
    }

    deinit {
        typingTimer?.invalidate()
    }

    // MARK: - Compute

    /// Чат E2E если личный и у обоих участников есть publicKey.
    var isE2E: Bool {
        guard chat.type == .direct else { return false }
        return chat.members.allSatisfy { $0.user.publicKey != nil && !($0.user.publicKey ?? "").isEmpty }
    }

    var otherMember: ChatMember? {
        chat.members.first(where: { $0.userId != currentUserId })
    }

    var displayName: String {
        chat.displayName(currentUserId: currentUserId)
    }

    // MARK: - Load

    func load() async {
        loading = true
        defer { loading = false }
        do {
            let result: [Message] = try await APIClient.shared.get("chats/\(chat.id)/messages")
            self.messages = result.sorted { $0.createdAt < $1.createdAt }
        } catch {
            self.error = (error as? APIError)?.errorDescription ?? error.localizedDescription
        }
    }

    // MARK: - Decryption helper

    /// Если сообщение зашифровано — возвращает расшифрованный текст или nil.
    /// Если plaintext — просто content.
    func displayContent(for message: Message) -> String? {
        guard let content = message.content else { return nil }
        if message.encrypted != true { return content }

        guard let nonce = message.nonce,
              let myPriv = privateKey
        else { return nil }

        // Для расшифровки берём публичный ключ другого участника
        // (E2E симметрична: ECDH(my_priv, their_pub) == ECDH(their_priv, my_pub)).
        let theirPub: String?
        if let senderId = message.senderId, senderId == currentUserId {
            theirPub = otherMember?.user.publicKey
        } else {
            theirPub = message.sender?.publicKey ?? otherMember?.user.publicKey
        }
        guard let theirPub else { return nil }

        return try? E2E.decryptText(
            ciphertextB64: content,
            nonceB64: nonce,
            senderPublicKeyB64: theirPub,
            recipientPrivateKeyB64: myPriv
        )
    }

    // MARK: - Send

    func sendDraft() {
        let text = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        draft = ""
        stopTypingImmediate()

        // ── Режим редактирования ──
        if let editing = editingMessage {
            sendEdit(messageId: editing.id, newText: text)
            editingMessage = nil
            return
        }

        let replyToId = replyingTo?.id
        replyingTo = nil

        let payload: SendMessagePayload
        if isE2E,
           let myPriv = privateKey,
           let theirPub = otherMember?.user.publicKey,
           let enc = try? E2E.encryptText(text, recipientPublicKeyB64: theirPub, senderPrivateKeyB64: myPriv) {
            payload = SendMessagePayload(
                chatId: chat.id,
                type: MessageType.text.rawValue,
                content: enc.ciphertextB64,
                nonce: enc.nonceB64,
                encrypted: true,
                mediaData: nil,
                replyToId: replyToId,
                forwardedFromId: nil
            )
        } else {
            payload = SendMessagePayload(
                chatId: chat.id,
                type: MessageType.text.rawValue,
                content: text,
                nonce: nil,
                encrypted: false,
                mediaData: nil,
                replyToId: replyToId,
                forwardedFromId: nil
            )
        }
        SocketClient.shared.send(payload)
    }

    // MARK: - Edit

    /// Начать редактировать (вызывается из контекстного меню).
    func beginEdit(_ message: Message) {
        guard message.senderId == currentUserId else { return }
        editingMessage = message
        replyingTo = nil
        // Заполняем draft уже-расшифрованным текстом
        draft = displayContent(for: message) ?? message.content ?? ""
    }

    func cancelEdit() {
        editingMessage = nil
        draft = ""
    }

    private func sendEdit(messageId: String, newText: String) {
        // Если чат E2E — шифруем новое содержимое тем же способом
        if isE2E,
           let myPriv = privateKey,
           let theirPub = otherMember?.user.publicKey,
           let enc = try? E2E.encryptText(newText, recipientPublicKeyB64: theirPub, senderPrivateKeyB64: myPriv) {
            SocketClient.shared.editMessage(
                messageId: messageId, chatId: chat.id,
                content: enc.ciphertextB64, nonce: enc.nonceB64, encrypted: true
            )
        } else {
            SocketClient.shared.editMessage(
                messageId: messageId, chatId: chat.id,
                content: newText, encrypted: false
            )
        }
    }

    // MARK: - Reply

    func beginReply(_ message: Message) {
        replyingTo = message
        editingMessage = nil
    }

    func cancelReply() {
        replyingTo = nil
    }

    // MARK: - Forward

    /// Пересылает сообщение в другой чат через тот же socket.io event.
    func forward(message: Message, to targetChatId: String) {
        let payload = SendMessagePayload(
            chatId: targetChatId,
            type: message.type.rawValue,
            content: nil,         // backend resolves from forwardedFromId
            nonce: nil,
            encrypted: false,
            mediaData: nil,
            replyToId: nil,
            forwardedFromId: message.id
        )
        SocketClient.shared.send(payload)
    }

    // MARK: - Image message

    func sendImage(prepared: ImagePreparer.Prepared) async {
        do {
            let media = try await MediaService.uploadImage(
                fileURL: prepared.fileURL,
                mimeType: prepared.mimeType,
                fileName: prepared.fileName
            )
            let payload = SendMessagePayload(
                chatId: chat.id,
                type: MessageType.image.rawValue,
                content: nil,
                nonce: nil,
                encrypted: false,
                mediaData: MediaPayload(
                    url: media.url,
                    mimeType: media.mimeType,
                    size: media.size,
                    duration: nil,
                    width: media.width ?? prepared.width,
                    height: media.height ?? prepared.height,
                    waveform: nil,
                    thumbnailUrl: nil
                ),
                replyToId: nil
            )
            SocketClient.shared.send(payload)
            try? FileManager.default.removeItem(at: prepared.fileURL)
        } catch {
            self.error = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
        }
    }

    // MARK: - Voice message

    /// Загружает голосовой файл и отправляет сообщение.
    func sendVoice(fileURL: URL, duration: TimeInterval, waveform: [CGFloat]) async {
        do {
            let media = try await MediaService.uploadVoice(
                fileURL: fileURL,
                duration: duration,
                waveform: waveform
            )
            let payload = SendMessagePayload(
                chatId: chat.id,
                type: MessageType.voice.rawValue,
                content: nil,
                nonce: nil,
                encrypted: false,
                mediaData: MediaPayload(
                    url: media.url,
                    mimeType: media.mimeType,
                    size: media.size,
                    duration: duration,
                    width: nil,
                    height: nil,
                    waveform: waveform.map { Double($0) },
                    thumbnailUrl: nil
                ),
                replyToId: nil
            )
            SocketClient.shared.send(payload)
            // Удаляем локальный временный файл
            try? FileManager.default.removeItem(at: fileURL)
        } catch {
            self.error = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
        }
    }

    // MARK: - Reactions / delete

    func react(messageId: String, emoji: String) {
        SocketClient.shared.reactToMessage(messageId: messageId, chatId: chat.id, emoji: emoji)
    }

    func delete(messageId: String) {
        SocketClient.shared.deleteMessageRemote(messageId: messageId, chatId: chat.id)
        // Локально удаляем сразу для отзывчивости (сервер придёт по сокету)
        messages.removeAll { $0.id == messageId }
    }

    // MARK: - Typing indicator

    func onTextChange() {
        SocketClient.shared.setTyping(chatId: chat.id, isTyping: true)
        typingTimer?.invalidate()
        typingTimer = Timer.scheduledTimer(withTimeInterval: 2, repeats: false) { [weak self] _ in
            Task { @MainActor in
                guard let self else { return }
                SocketClient.shared.setTyping(chatId: self.chat.id, isTyping: false)
            }
        }
    }

    private func stopTypingImmediate() {
        SocketClient.shared.setTyping(chatId: chat.id, isTyping: false)
        typingTimer?.invalidate()
        typingTimer = nil
    }

    // MARK: - Socket subscriptions

    private func subscribeToSocket() {
        SocketClient.shared.messageReceived
            .filter { [weak self] in $0.chatId == self?.chat.id }
            .sink { [weak self] msg in
                guard let self else { return }
                // Не дублируем (на случай оптимистичной локальной вставки)
                guard !self.messages.contains(where: { $0.id == msg.id }) else { return }
                self.messages.append(msg)
                // Если входящее — сразу помечаем прочитанным
                if msg.senderId != self.currentUserId {
                    SocketClient.shared.markRead(messageId: msg.id, chatId: self.chat.id)
                }
            }
            .store(in: &cancellables)

        SocketClient.shared.messageDeleted
            .filter { [weak self] in $0.chatId == self?.chat.id }
            .sink { [weak self] event in
                self?.messages.removeAll { $0.id == event.messageId }
            }
            .store(in: &cancellables)

        SocketClient.shared.typingChanged
            .filter { [weak self] in $0.chatId == self?.chat.id && $0.userId != self?.currentUserId }
            .sink { [weak self] event in
                guard let self else { return }
                if event.isTyping { self.typingUsers.insert(event.userId) }
                else              { self.typingUsers.remove(event.userId) }
            }
            .store(in: &cancellables)

        SocketClient.shared.messageUpdated
            .sink { [weak self] event in
                guard let self,
                      let idx = self.messages.firstIndex(where: { $0.id == event.id }) else { return }
                // Поскольку Message — value type с let, делаем полный re-fetch текущего чата.
                // Это просто и надёжно; для оптимизации можно ввести мутабельные filed-views.
                _ = idx
                Task { await self.load() }
            }
            .store(in: &cancellables)
    }
}

File: apps/android/app/src/main/java/online/akkdmsg/dakka/chat/ChatViewModel.kt
package online.akkdmsg.dakka.chat

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import online.akkdmsg.dakka.auth.AuthStore
import online.akkdmsg.dakka.data.Chat
import online.akkdmsg.dakka.data.MediaPayload
import online.akkdmsg.dakka.data.Message
import online.akkdmsg.dakka.data.SendMessagePayload
import online.akkdmsg.dakka.data.api.ChatApi
import online.akkdmsg.dakka.data.api.MediaService
import online.akkdmsg.dakka.data.api.SocketClient
import online.akkdmsg.dakka.data.crypto.E2E
import online.akkdmsg.dakka.data.isE2E
import online.akkdmsg.dakka.data.otherMember
import online.akkdmsg.dakka.media.ImagePreparer
import java.io.File

class ChatViewModel(
    private val chat: Chat,
    private val currentUserId: String,
    private val privateKey: String?,
) : ViewModel() {

    val chatRef: Chat = chat

    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    private val _loading = MutableStateFlow(false)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    private val _draft = MutableStateFlow("")
    val draft: StateFlow<String> = _draft.asStateFlow()

    private val _typingUsers = MutableStateFlow<Set<String>>(emptySet())
    val typingUsers: StateFlow<Set<String>> = _typingUsers.asStateFlow()

    private val _replyingTo = MutableStateFlow<Message?>(null)
    val replyingTo: StateFlow<Message?> = _replyingTo.asStateFlow()

    private val _editing = MutableStateFlow<Message?>(null)
    val editing: StateFlow<Message?> = _editing.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    val isE2E: Boolean get() = chat.isE2E
    val peerName: String get() = chat.otherMember(currentUserId)?.user?.displayName ?: chat.name.orEmpty()
    val peerAvatar: String? get() = chat.otherMember(currentUserId)?.user?.avatar ?: chat.avatar

    private var typingTimer: Job? = null

    init {
        load()
        subscribeToSocket()
    }

    fun setDraft(s: String) {
        _draft.value = s
        onTextChange()
    }

    private fun load() {
        viewModelScope.launch {
            _loading.value = true
            try {
                _messages.value = ChatApi.listMessages(chat.id)
                    .sortedBy { it.createdAt }
            } catch (e: Exception) {
                _error.value = e.message ?: "Не удалось загрузить сообщения"
            } finally {
                _loading.value = false
            }
        }
    }

    fun sendDraft() {
        val t = _draft.value.trim()
        if (t.isEmpty()) return
        SocketClient.setTyping(chat.id, false)

        // Edit mode: отправляем message:edit вместо нового сообщения
        val editingMsg = _editing.value
        if (editingMsg != null) {
            if (isE2E) {
                // SECURITY: при E2E-чате edit ОБЯЗАН быть зашифрован. Никакого
                // silent plaintext fallback.
                val recipient = chat.otherMember(currentUserId)
                val recipientPub = recipient?.user?.publicKey
                if (privateKey.isNullOrBlank() || recipientPub.isNullOrBlank()) {
                    _error.value = "Не удалось зашифровать: ключ недоступен. Перелогиньтесь."
                    return
                }
                val enc = try {
                    E2E.encryptText(t, recipientPub, privateKey)
                } catch (e: Exception) {
                    _error.value = "Ошибка шифрования. Сообщение не изменено."
                    return
                }
                _editing.value = null
                _draft.value = ""
                SocketClient.editMessage(editingMsg.id, chat.id, enc.ciphertextB64, enc.nonceB64, true)
            } else {
                _editing.value = null
                _draft.value = ""
                SocketClient.editMessage(editingMsg.id, chat.id, t, null, false)
            }
            return
        }

        val replyId = _replyingTo.value?.id
        val recipient = chat.otherMember(currentUserId)
        val recipientPub = recipient?.user?.publicKey

        val payload: SendMessagePayload = if (isE2E) {
            // SECURITY: E2E-чат ОБЯЗАН быть зашифрован. Никакого silent plaintext.
            if (privateKey.isNullOrBlank()) {
                _error.value = "Не удалось отправить: приватный ключ недоступен. Перелогиньтесь."
                return
            }
            if (recipientPub.isNullOrBlank()) {
                _error.value = "У получателя нет ключа шифрования."
                return
            }
            val enc = try {
                E2E.encryptText(t, recipientPub, privateKey)
            } catch (e: Exception) {
                _error.value = "Ошибка шифрования. Сообщение не отправлено."
                return
            }
            _draft.value = ""
            _replyingTo.value = null
            SendMessagePayload(
                chatId = chat.id, type = "text",
                content = enc.ciphertextB64, nonce = enc.nonceB64,
                encrypted = true, replyToId = replyId,
            )
        } else {
            _draft.value = ""
            _replyingTo.value = null
            SendMessagePayload(chatId = chat.id, type = "text", content = t,
                encrypted = false, replyToId = replyId)
        }
        SocketClient.send(payload)
    }

    // ── Reply / Edit / Delete / React / Forward ──────────────────────────────

    fun beginReply(message: Message) {
        _replyingTo.value = message
        _editing.value = null
    }
    fun cancelReply() { _replyingTo.value = null }

    fun beginEdit(message: Message) {
        if (message.senderId != currentUserId) return
        _editing.value = message
        _replyingTo.value = null
        _draft.value = displayContent(message) ?: message.content.orEmpty()
    }
    fun cancelEdit() {
        _editing.value = null
        _draft.value = ""
    }

    fun react(message: Message, emoji: String) {
        SocketClient.reactToMessage(message.id, chat.id, emoji)
    }

    fun delete(message: Message) {
        SocketClient.deleteMessageRemote(message.id, chat.id)
        // Локально удаляем для отзывчивости
        _messages.value = _messages.value.filter { it.id != message.id }
    }

    fun forward(message: Message, targetChatId: String) {
        SocketClient.send(SendMessagePayload(
            chatId = targetChatId,
            type = message.type,
            forwardedFromId = message.id,
        ))
    }

    /** Короткое превью для ReplyBar/cited-message. */
    fun previewText(msg: Message): String = when (msg.type) {
        "voice"  -> "🎙 Голосовое"
        "image"  -> "📷 Фото"
        "circle" -> "⭕ Видео-кружок"
        "video"  -> "🎬 Видео"
        "file"   -> "📎 Файл"
        else     -> displayContent(msg)?.takeIf { it.isNotBlank() } ?: msg.content.orEmpty()
    }

    private val _scrollTarget = MutableStateFlow<String?>(null)
    val scrollTarget: StateFlow<String?> = _scrollTarget.asStateFlow()
    fun requestScrollTo(messageId: String) { _scrollTarget.value = messageId }
    fun consumeScrollTarget() { _scrollTarget.value = null }

    /**
     * Возвращает текст для отображения: расшифрованный (если encrypted),
     * иначе plaintext content, иначе null.
     */
    fun displayContent(msg: Message): String? {
        val raw = msg.content ?: return null
        if (msg.encrypted != true) return raw

        val nonce = msg.nonce ?: return null
        val myPriv = privateKey ?: return null

        // E2E симметрично: ECDH(my_priv, their_pub) даёт тот же AES key для обоих
        val theirPub = if (msg.senderId == currentUserId) {
            chat.otherMember(currentUserId)?.user?.publicKey
        } else {
            msg.sender?.publicKey ?: chat.otherMember(currentUserId)?.user?.publicKey
        } ?: return null

        return try {
            E2E.decryptText(
                ciphertextB64 = raw,
                nonceB64 = nonce,
                senderPublicKeyB64 = theirPub,
                recipientPrivateKeyB64 = myPriv,
            )
        } catch (_: Exception) {
            null
        }
    }

    // ── Голос ────────────────────────────────────────────────────────────────

    fun sendVoice(file: File, durationSec: Int, waveform: List<Float>) {
        viewModelScope.launch {
            try {
                val media = MediaService.uploadVoice(file, durationSec, waveform)
                val payload = SendMessagePayload(
                    chatId = chat.id,
                    type = "voice",
                    mediaData = MediaPayload(
                        url = media.url,
                        mimeType = media.mimeType,
                        size = media.size,
                        duration = durationSec.toDouble(),
                        waveform = waveform.map { it.toDouble() },
                    ),
                )
                SocketClient.send(payload)
                file.delete()
            } catch (e: Exception) {
                _error.value = "Не удалось отправить голосовое: ${e.message}"
            }
        }
    }

    // ── Картинка ─────────────────────────────────────────────────────────────

    fun sendImage(context: android.content.Context, uri: android.net.Uri) {
        viewModelScope.launch {
            try {
                val prepared = ImagePreparer.fromUri(context, uri) ?: return@launch
                val media = MediaService.uploadImage(prepared.file, prepared.mimeType, prepared.fileName)
                val payload = SendMessagePayload(
                    chatId = chat.id,
                    type = "image",
                    mediaData = MediaPayload(
                        url = media.url,
                        mimeType = media.mimeType,
                        size = media.size,
                        width = media.width ?: prepared.width,
                        height = media.height ?: prepared.height,
                    ),
                )
                SocketClient.send(payload)
                prepared.file.delete()
            } catch (e: Exception) {
                _error.value = "Не удалось отправить фото: ${e.message}"
            }
        }
    }

    // ── Видео (полноразмерное) ───────────────────────────────────────────────

    fun sendVideo(file: File, durationSec: Int) {
        viewModelScope.launch {
            try {
                if (file.length() > 100L * 1024 * 1024) {
                    _error.value = "Видео слишком большое (макс. 100 МБ)"
                    file.delete()
                    return@launch
                }
                val media = MediaService.uploadVideo(file, durationSec)
                val payload = SendMessagePayload(
                    chatId = chat.id,
                    type = "video",
                    mediaData = MediaPayload(
                        url = media.url,
                        mimeType = media.mimeType,
                        size = media.size,
                        duration = durationSec.toDouble(),
                        width = media.width,
                        height = media.height,
                    ),
                )
                SocketClient.send(payload)
                file.delete()
            } catch (e: Exception) {
                _error.value = "Не удалось отправить видео: ${e.message}"
            }
        }
    }

    /**
     * Загрузить готовый видео-файл из системного picker'а (content:// uri).
     * Копирует во временный кэш-файл и шлёт через [sendVideo].
     */
    fun sendVideoFromUri(context: android.content.Context, uri: android.net.Uri) {
        viewModelScope.launch {
            try {
                val resolver = context.contentResolver
                val tmp = File(context.cacheDir, "video-pick-${System.currentTimeMillis()}.mp4")
                resolver.openInputStream(uri)?.use { input ->
                    tmp.outputStream().use { out -> input.copyTo(out) }
                } ?: run {
                    _error.value = "Не удалось прочитать видео"
                    return@launch
                }
                if (tmp.length() > 100L * 1024 * 1024) {
                    _error.value = "Видео слишком большое (макс. 100 МБ)"
                    tmp.delete()
                    return@launch
                }
                // duration не известен из uri — оставим 0, сервер всё равно сохранит как есть
                sendVideo(tmp, 0)
            } catch (e: Exception) {
                _error.value = "Не удалось отправить видео: ${e.message}"
            }
        }
    }

    // ── Кружок ───────────────────────────────────────────────────────────────

    fun sendCircle(file: File, durationSec: Int) {
        viewModelScope.launch {
            try {
                val media = MediaService.uploadCircle(file, durationSec)
                val payload = SendMessagePayload(
                    chatId = chat.id,
                    type = "circle",
                    mediaData = MediaPayload(
                        url = media.url,
                        mimeType = media.mimeType,
                        size = media.size,
                        duration = durationSec.toDouble(),
                        width = media.width,
                        height = media.height,
                    ),
                )
                SocketClient.send(payload)
                file.delete()
            } catch (e: Exception) {
                _error.value = "Не удалось отправить кружок: ${e.message}"
            }
        }
    }

    // ── Typing indicator с 2-секундным дебаунсом ──────────────────────────────

    private fun onTextChange() {
        SocketClient.setTyping(chat.id, true)
        typingTimer?.cancel()
        typingTimer = viewModelScope.launch {
            delay(2000)
            SocketClient.setTyping(chat.id, false)
        }
    }

    // ── Socket subscriptions ──────────────────────────────────────────────────

    private fun subscribeToSocket() {
        viewModelScope.launch {
            SocketClient.messageReceived.collect { msg ->
                if (msg.chatId != chat.id) return@collect
                if (_messages.value.any { it.id == msg.id }) return@collect
                _messages.value = _messages.value + msg
                // авто-mark-read если входящее
                if (msg.senderId != currentUserId) {
                    SocketClient.markRead(msg.id, chat.id)
                }
            }
        }
        viewModelScope.launch {
            SocketClient.messageDeleted.collect { event ->
                if (event.chatId != chat.id) return@collect
                _messages.value = _messages.value.filter { it.id != event.messageId }
            }
        }
        viewModelScope.launch {
            SocketClient.typingChanged.collect { event ->
                if (event.chatId != chat.id || event.userId == currentUserId) return@collect
                _typingUsers.value = if (event.isTyping) _typingUsers.value + event.userId
                                    else _typingUsers.value - event.userId
            }
        }
        // Реакции, edit — бэк шлёт только messageId; перезатягиваем список.
        viewModelScope.launch {
            SocketClient.messageUpdated.collect { messageId ->
                if (!_messages.value.any { it.id == messageId }) return@collect
                try {
                    _messages.value = ChatApi.listMessages(chat.id).sortedBy { it.createdAt }
                } catch (_: Exception) { /* offline-ok */ }
            }
        }
    }
}
